Operator splitting

jdannberg opened the pull request:
This pull request adds the option to use operator splitting as a solver scheme to solve advection and reactions between compositional fields (and temperature) separately. For now, there is only one solver scheme that uses this, and I've only implemented a simple first-order operator split that uses a forward Euler as a solver. These things could be improved in the future. 
However, I've also added a benchmark for this to the manual, and for the cases I tested, the error caused by the operator split was usually small, and the time used to compute the reactions was also small (compared to the advection). 

Everything seems to work, but I still have to work on the documentation, so I've marked it as work in progress.

tjhei commented on benchmarks/operator_splitting/advection_reaction/advection_reaction.base.prm:
```diff
@@ -0,0 +1,162 @@
+# In comparison to the example 'exponential_decay', the
+# initial comosition depends on x, and there is a constant 
+# horizontal velocity. The composition decays over time, 
+# following the equations for exponetial decay with a decay 
```
typo "exponential"

tjhei commented on include/aspect/heating_model/interface.h:
```diff
@@ -64,14 +64,23 @@ namespace aspect
+       * at the given position. This includes for example latent heat of
+       * melt.
+       */
+      std::vector<double> heating_reaction_terms;
```
Are these ignored if the solver scheme is _not_ operator-split? Can we also account for them instead?

jdannberg commented:
Okay, I updated the pull request. 

> there are going to be quite a few different variations of the operator split we might be interested in doing (nonlinear Stokes yes/no, different time stepping schemes, different splitting schemes, ...). Would you suggest to have them as additional options within the "operator-split" scheme? Otherwise we would need to change the name to something more unique like "forward-euler-single-Stokes-solve-operator-split". ;-)

I think for now, they could be options that can be added to the "operator split" scheme. As we discussed restructuring the solver schemes at some point in the near future anyway (so that solvers can be mixed and matched) I did not spend much time on this in this pull request. 

> Can you explain how the new additional heating term outputs work with the old vs. the new scheme? We now have "reaction terms" and "source terms", are both usable whether you use operator split or not?

The way this is designed right now is that operator splitting uses both "reaction terms" and "source terms" outputs, and the old scheme stays exactly as it was before (it only uses the "old" terms). This means that it is the responsibility of the material or heating model, if it wants to be compatible with operator splitting, to fill both the "reaction terms" and "source terms" in a way that work with both solver schemes. I adapted 3 material models and one heating model to do this as part of this pull request, and they basically now fill the "old" output if there is no operator splitting, and fill the "new" output in case of operator splitting, and the respective other term is set to zero. I added some more documentation, so I hope this becomes clear now. 
The reason why I do it like this is that my new terms are not part of an assembler, so I never get to decide if I would want to leave out some terms of the assembly (unless I would add if-statements to each assembler, asking if the solver scheme is "operator splitting", which would not be great). This also means that I do not need to make any changes to the assembly, the new terms are only used in the new solver scheme. 

> have you tried this new scheme with a melting example?

I added a test with melt transport (latent_heat_melt_transport) and it looks reasonable. 

There are a few more things:
I had to change the interface of the `NamedAdditionalMaterialOutputs` class (which we just added at the hackathon), namely the `get_nth_output` function. It used to return a reference to the nth output instead of the vector itself, but that does not work if this vector is only created locally. In my case (`ReactionRateOutputs`), the outputs are a `std::vector<std::vector<double> > reaction_rates`, but `reaction_rates.size()` has to return the number of evaluation points (as opposed to the number of compositional fields), because this is the way we use this structure in the material model. This means that I can not just return `reaction_rates[nth_compositional_field]` from `get_nth_output`, but I would have to sort the entries in my vector in a different way. But I can not just create a new vector locally, because I have to return a reference. To fix this, I changed the interface to return the vector instead of a reference. I could also move this to a separate pull request, if that seems more appropriate. 

Finally, a question: In some places, I have to check if the reaction rate terms have been filled by the material model. For example, someone could use a heating model that is based on getting the reaction rates from the material model outputs (as is the case for the `latent_heat_melt` heating model), so I have to check if they are actually there. If I initialize the reaction rates with signaling_NaNs, is there any way for me to check if they are NaN or not? At the moment, I initialize them with quiet_NaNs, because I couldn't figure out a way to do anything with them if they were signaling_NaNs (I always triggered a floating point exception rather than my Assert).

jdannberg commented:
> there are going to be quite a few different variations of the operator split we might be interested in doing (nonlinear Stokes yes/no, different time stepping schemes, different splitting schemes, ...). Would you suggest to have them as additional options within the "operator-split" scheme? Otherwise we would need to change the name to something more unique like "forward-euler-single-Stokes-solve-operator-split". ;-)

Alternatively, I had a different idea for that. If we stick with first-order operator split, we can just make the operator split a solver option instead of a new solver scheme. We could add it in all of the existing schemes in the beginning, before the main loop where the iterations are done (in a similar way as it is done for the free surface right now), and it would only be two lines of code in each of those places (we just have to call the compute_reactions() function).
That would save a lot of code duplication, it would work for all solver schemes right from the start, and I tested it with a new melt cookbook that I am writing and it seems to work quite well.

bangerth commented:
That would also work if you did a half-reaction step first, then the regular step, then another half-reaction step -- that's one of the two ways of doing the second order (Strang) splitting.

jdannberg commented:
Yes, good point!

tjhei commented:
> The way this is designed right now is that operator splitting uses both "reaction terms" and "source terms" outputs, and the old scheme stays exactly as it was before (it only uses the "old" terms). This means that it is the responsibility of the material or heating model, if it wants to be compatible with operator splitting, to fill both the "reaction terms" and "source terms" in a way that work with both solver schemes.

So this means that:
if operator-split if off, both contributions are inside "source terms", but if it is turned on, they are split between "reaction" and "source terms"? I don't like that this logic then needs to be in every plugin. Wouldn't it be easier to assemble ``dt*reaction_terms`` if operator-split if off?

tjhei commented:
> I had to change the interface of the NamedAdditionalMaterialOutputs class (which we just added at the hackathon), namely the get_nth_output function. It used to return a reference to the nth output instead of the vector itself, but that does not work if this vector is only created locally.

Not particularly pretty. What do you think about swapping the order you index the outputs? I think we keep rearranging these arrays for historical reasons mostly.

jdannberg commented:
> So this means that:
if operator-split if off, both contributions are inside "source terms", but if it is turned on, they are split between "reaction" and "source terms"? I don't like that this logic then needs to be in every plugin. Wouldn't it be easier to assemble dt*reaction_terms if operator-split if off?

Right now, the logic only has to be in the plugins that describe reactions (so there are not that many). 
I think we talked about this before and the reason we implemented it like this was that we didn't want the assembly to have if-statements asking for the solver scheme (and I guess they would also need to be added to the `compute_advection_system_residual` function). We would need them for both temperature and composition, and we would also have to check if the material model even fills these outputs, so we would make the assembly more complex than it already is. But if you think it is better to have that logic in the assembly, it should not be that much effort to change it.

jdannberg commented:
> Not particularly pretty. What do you think about swapping the order you index the outputs? I think we keep rearranging these arrays for historical reasons mostly.

That's what I started with, but then you would have different indexing in the `reaction_terms` and the `reaction_rates`. It would be `out.reaction_terms[point][composition]`, and `reaction_out->reaction_rates[composition][point]`. I found that very confusing and a potential source of errors, so I thought what I have right now makes more sense. Plus, the material model usually assumes that if you ask for the size of an output, it will be the number of evaluation points, which wouldn't work any more for the reaction rates.

jdannberg commented:
This addresses #1542 (just wanted to reference it here, so that we don't forget to close the issue).

tjhei commented:
> but then you would have different indexing in the reaction_terms and the reaction_rates

I agree, that is terrible. Having to allocate and move a vector every time you use ``get_nth_output`` (so on every cell) is not great though. The only other option I could think of is to pass in a reference and having it filled inside. Not sure if that is much better though.

jdannberg commented:
> The only other option I could think of is to pass in a reference and having it filled inside. Not sure if that is much better though.

Hm. Anyone else having any ideas for this? I already played around with this for some time and couldn't come up with anything better...

bangerth commented:
I'm not excited about the material models having to put stuff into two different places, depending on whether or not we use operator splitting. My preference would be to keep the user-facing side of the code base (i.e., the plugins) as simple as we can, and instead move the necessary `if` statements to the assembly side.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -9684,6 +9684,104 @@ \subsubsection{The solitary wave benchmark}
+
+Instead of solving the coupled equation
+\begin{align}
+  \frac{\partial c(t)}{\partial t} + \mathbf u\cdot\nabla c(t)
```
here and below, whenever we refer to *all* compositions, we use `\mathfrak{c}`.

jdannberg replied:
Okay, I hope I got all of them...

bangerth commented on include/aspect/heating_model/interface.h:
```diff
@@ -64,14 +64,36 @@ namespace aspect
+       * In contrast to the heating source terms, these terms are actual changes
+       * in temperature (units K/s or K/yr) rather than changes in energy.
+       */
+      std::vector<double> rates_of_temperature_change;
```
As mentioned in a previous comment, I'm not excited about having two fields with a similar meaning in this structure, in particular since they are not both used by all schemes. Can we find a way so that the material model doesn't have to know what solver is used?

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -632,7 +632,7 @@ namespace aspect
          * values of the additional output with that index.
          */
-        virtual const std::vector<double> &get_nth_output(const unsigned int idx) const = 0;
+        virtual const std::vector<double> get_nth_output(const unsigned int idx) const = 0;
```
If you return the object by value, there is no reason to mark it as `const`.

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -655,7 +655,7 @@ namespace aspect
         SeismicAdditionalOutputs(const unsigned int n_points);
 
-        virtual const std::vector<double> &get_nth_output(const unsigned int idx) const;
+        virtual const std::vector<double> get_nth_output(const unsigned int idx) const;
```
same here

bangerth commented on include/aspect/parameters.h:
```diff
@@ -307,6 +307,8 @@ namespace aspect
     double                         CFL_number;
     double                         maximum_time_step;
+    double                         reaction_time_step;
+    int                            reaction_steps_per_advection_step;
```
I think we should make these local to the ODE integrator if we can. In ODE schemes, the time step size can often be chosen automatically. For the moment, just choose it somehow reasonable as a local variable. Or do you actually set it in input files?

jdannberg replied:
At the moment, I need it for the benchmark cases, because to demonstrate that the scheme works correctly, I vary the `reaction_time_step` and show convergence (in case you haven't looked at the pdf files yet -- github doesn't show them by default -- they are all convergence plots).

bangerth commented on include/aspect/simulator.h:
```diff
@@ -924,6 +924,30 @@ namespace aspect
+       *
+       * As the ordinary differential equation in any given point is independent
+       * from the solution at all other points, we do not have to assemble a matrix,
+       * but just need to loop over all points and compute the update to the solution.
```
points -> node locations for the temperature and compositional fields

bangerth commented on include/aspect/simulator.h:
```diff
@@ -924,6 +924,30 @@ namespace aspect
+       * but just need to loop over all points and compute the update to the solution.
+       *
+       * The function also updates the old solution with the reaction update so that
+       * the advection scheme will have the correct time stepping in the next step.
```
time stepping? do you mean something like "starting guess"? (I think I know what you mean from our discussion of what the old solutions need to store, but it may not be obvious to a reader).

jdannberg replied:
No, I really mean the time stepping. Imagine the simple case where the velocity is zero, there is no diffusion, and the assembled matrix for the advection also does not include any reaction terms, because those are handled by the operator split. Then, in the first time step, all that is assembled is basically `scratch.phi_field[i] * scratch.phi_field[j] * JxW`on the left-hand side, and `scratch.old_field_values[q] * scratch.phi_field[i]` on the right-hand side, so basically solution = old_solution. But we want that to be the old_solution that already includes the reactions, otherwise our new solution doesn't make any sense. 
I updated the description to be more precise.

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1484,6 +1484,212 @@ namespace aspect
+  void Simulator<dim>::compute_reactions ()
+  {
+    if (time_step == 0)
+      return;
```
can you add why this works? (I suspect it's because the first time step length is zero, right? that could be checked by an assertion.)

jdannberg replied:
`time_step` already is the time step length (the time step number is called `timestep_number`). I was confused about that myself several times already. I'll add a comment.

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1484,6 +1484,212 @@ namespace aspect
+          in_C.reinit(fe_values_C, &cell, introspection, solution);
+
+          fe_values_T.reinit (cell);
+          cell->get_dof_indices (local_dof_indices);
```
you already called this in line 1563 above

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1484,6 +1484,212 @@ namespace aspect
+
+                // skip entries that are not locally owned:
+                if (!dof_handler.locally_owned_dofs().is_element(local_dof_indices[composition_idx]))
+                  continue;
```
just revert the condition and put the following two lines into `{...}`. I think that would be easier to read.

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1484,6 +1484,212 @@ namespace aspect
+
+                // skip entries that are not locally owned:
+                if (!dof_handler.locally_owned_dofs().is_element(local_dof_indices[temperature_idx]))
+                  continue;
```
same here

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -1484,6 +1484,212 @@ namespace aspect
+                  continue;
+
+                distributed_vector(local_dof_indices[composition_idx]) = in_C.composition[j][c];
+                distributed_reaction_vector(local_dof_indices[composition_idx]) = accumulated_reactions_C[j][c];
```
here, and in the corresponding code below: what happens for DoFs that are located on the interfaces between cells? don't we add up twice the correct update because we start from the wrong starting point the second time around?

jdannberg replied:
I don't know, I just stole this from the `MeltHandler<dim>::compute_melt_variables` function in melt.cc and there it is the same. Maybe @tjhei can comment on this?

tjhei replied:
This is a long standing war I have with Wolfgang, I prefer ``continue``, he prefers nested blocks. We have to fight that one out eventually. ;-)

bangerth replied:
Let us not choose bikes. Maybe a coding battle? ;-)

bangerth replied:
But, @tjhei, your comment is not on point here. The question wasn't about the `continue` but about whether we are setting the same DoFs twice. I can't convince myself that we're not adding contributions twice here.

tjhei replied:
Oh, I meant to post this below your comment above.

> whether we are setting the same DoFs twice

Correct, this will happen. I think the way we originally designed the algorithm in Georgia we accounted for that (correct result but computing the ODEs more often than necessary). I am not sure how it is working now.

jdannberg replied:
I have to admit I don't really understand what the problem is. Do you think the code is wrong? Or do you think it does unnecessary work? 
Because I thought we're only assigning values (not adding up things), so why does it matter if we set the same DoF twice? Also, shouldn't that be visible in my benchmark results if those DoFs had the wrong values? What do you suggest?

And I don't think I changed anything in that algorithm we designed in Georgia.

tjhei replied:
Values are written more than once. Whether this leads to incorrect results, I don't know because I haven't carefully checked the code. If the computation reads in the updated values, things are likely incorrect. If you don't read from them, things are probably correct.
In any case, we are duplicating quite a bit of work. It would make sense to figure out if we can remove this overhead. I am not sure how to do that though.

bangerth replied:
I think I have convinced myself that the code is indeed correct. We initialize `in_T` and `in_C` from the `solution` vector on every cell, and on every cell put the result into the `distributed_vector` vector. Only after the loop over all cells do we actually copy `distributed_vector` back onto `solution`. So touching the same DoF twice starts at the same value, computes the same value, and overwrites the same value in `distributed_vector`. It does not start at the updated value already. Maybe something like that ought to go into the documentation at the top of the function?

For a Q1 element in 3d, we will be doing the same work 8 times over (because each vertex is part of 8 cells). That should be easy to avoid, but can wait for a follow-up patch.

tjhei replied:
Yes, a comment in front of that loop would be good (with the important part that we never read from ``distributed_vector`` inside the loop). I am not sure what your "easy" approach is because I can only think of keeping track of already updated values using a bit vector. @jdannberg can you create an issue for this?

bangerth replied:
Yes, I was thinking of a bit vector or an `IndexSet`.

jdannberg replied:
Okay, I added the comment and opened a new issue for operator splitting with a list of potential improvements.

jdannberg commented:
I see that having these things in the material model is not nice, but what I noticed is that for the cases where the operator splitting is important (for example the melt models), you may want to fill two different things in the material model anyway. For example for the melting rate, right now we have to use the compositional fields from the old time step to compute the equilibrium melt fraction (because we do not want that to change in the nonlinear iterations). But if we use operator splitting, we can use the usual material models inputs to compute the equilibrium melt fraction (which in this case is also more accurate than the old solution). 
That means that exactly in those cases where you would want to use operator splitting, you would have to make the distinction in the material model anyway.

bangerth commented:
So what you're saying is that one field takes into account *instanteneous* changes (i.e., very fast compared to the advection time step) whereas the other one takes into account changes on time scales roughly equal or longer than the advection time step?

If that's correct, then we could just leave it to the material models to fill one or both of these fields, and both the current code and the operator-splitting code could just handle both. This way, there's less potential for getting things wrong. Existing code would also continue to work.

jdannberg commented:
> So what you're saying is that one field takes into account instanteneous changes (i.e., very fast compared to the advection time step) whereas the other one takes into account changes on time scales roughly equal or longer than the advection time step?

Yes, that is how I imagined it. 

> If that's correct, then we could just leave it to the material models to fill one or both of these fields, and both the current code and the operator-splitting code could just handle both. This way, there's less potential for getting things wrong. Existing code would also continue to work.

This is how it works at the moment, with the exception that the current code (without operator splitting) does not handle the instantaneous changes. How would you want to do that? We could add them in the assembly, but because they are instantaneous changes, they might not work correctly in models without operator split. 
I agree that it would be nice if we could make the distinction between the instantaneous and slow outputs more clear. But one problem is that the old reaction_terms were used for both things (instantaneous and slow changes), and that worked in cases where the problem was otherwise linear, or the reactions did not depend on the composition/temperature fields. For example, the "Using passive and active compositional fields" cookbook has a case with reactions that are essentially instantaneous. As soon as you use a nonlinear solver, that doesn't work correctly any more. Now we could say that these instantaneous changes are always put into the new (instantaneous) reaction_rates output, and just add it in the assembly (which would also mean that we would have if-statements asking if operator splitting is used in all versions of the advection field assembly). But then people might fill this new output in the material model, and if they do not use operator splitting, it would just not work correctly. For these material models that handle instantaneous changes, they should essentially always be used with operator splitting, and just for reasons of backwards compatibility they now have these if-statements and continue to fill the old reaction_terms outputs instead of only the new instantaneous ones. For a new material model handling reactions, I'd write it in a way that if you wanted to use the reactions, it would only work with operator splitting.

jdannberg commented:
So what do we do about the different outputs in the material model/heating model? I like it the way it is right now, but you both seemed to prefer if-statements in the assembly rather than assigning values to different outputs in the evaluate() function in the material model. Should I move these things to the assembly (which would have the drawbacks described in my long comment above), or do we leave it like it is right now?

tjhei commented:
> So what do we do about the different outputs in the material model/heating model?

Good question. I have another suggestion worth trying:
Can we add an assert that the new reaction_terms are 0.0 if operator split is disabled? I think this would help prevent mistakes where you fill the wrong entries.

jdannberg commented:
Okay, done. 
The only question: Do we really want to allow zero? By default, these outputs are not created at all, and the material models that do create them fill them with quietNans. So in the Assert, I allow them to be Nan, and I also allow them to be zero right now.

tjhei commented:
> Do we really want to allow zero? By default, these outputs are not created at all.

Oh, then I did not remember correctly. The idea was obviously to detect if somebody fills them even though they should not be filled. How that is done best I leave to you.

tjhei commented on source/heating_model/latent_heat_melt.cc:
```diff
@@ -36,25 +36,60 @@ namespace aspect
+                                                                  * melting_rate
+                                                                  * material_model_inputs.temperature[q];
+                }
+              else if (use_operator_split && reaction_rate_out != NULL)
```
shouldn't it be an internal error if ``use_operator_split`` but ``reaction_rate_out == NULL``? Maybe turn the second part into an assert then?

jdannberg replied:
But I think there are many places in the code where we evaluate the heating model, but do not call `create_additional_named_outputs` before, and in those cases reaction_rate_out == NULL (like in the normal assembly). That is a totally valid case. The entries in reaction_rate_out just have to be filled in case it exists, so they can't be NaNs (which is covered in the assert below), but if the `ReactionRateOutputs` have not been created, that probably just means that in the place where we are in the code they are not needed.

jdannberg replied:
Or now that I think about it, the assembly probably creates the additional named outputs, but for example the "stress" postprocessor only needs the viscosity, so it evaluates the material model without calling `create_additional_named_outputs`.

tjhei commented on source/heating_model/latent_heat_melt.cc:
```diff
@@ -36,25 +36,60 @@ namespace aspect
+                  // if operator splitting is used in the model, we have to use the reaction rates from the
+                  // material model outputs instead of the reaction terms
+                  AssertThrow (std::isfinite(reaction_rate_out->reaction_rates[q][porosity_idx]),
+                               ExcMessage ("You are trying to use reaction rate outputs from the material"
```
missing space at end.

tjhei commented on source/material_model/composition_reaction.cc:
```diff
@@ -90,6 +92,19 @@ namespace aspect
+              // (and then set the latter to zero).
+              if (this->get_parameters().use_operator_splitting)
+                {
+                  if (reaction_rate_out != NULL)
```
again, what happens if it is NULL? Shouldn't that be an error?

jdannberg replied:
See my reply above.

tjhei commented on source/material_model/melt_global.cc:
```diff
@@ -183,6 +191,21 @@ namespace aspect
+                  // fill reaction rate outputs if the model uses operator splitting
+                  if (this->get_parameters().use_operator_splitting)
+                    {
+                      if (reaction_rate_out != NULL)
```
assert instead?

tjhei commented on source/material_model/composition_reaction.cc:
```diff
@@ -243,6 +258,20 @@ namespace aspect
+    void
+    CompositionReaction<dim>::create_additional_named_outputs (MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      if (out.template get_additional_output<ReactionRateOutputs<dim> >() == NULL)
```
does this mean (here and in all other places) that you always create these outputs even when we don't use operator splitting? That doesn't seem right.

jdannberg replied:
Hm, I guess you are right. We now only create them if operator splitting is used.

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -1497,6 +1496,30 @@ namespace aspect
 
+    // make sure that if the model does not use operator splitting,
+    // the material model outputs do not fill the reaction_rates (because the reaction_terms are used instead)
+    if (!parameters.use_operator_splitting)
```
can you wrap the whole thing in
```
#ifdef DEBUG

#endif
```
?

tjhei replied:
what happened to this?

jdannberg replied:
Oh, sorry, I totally didn't see that last time. Should be fixed now.

tjhei commented on source/simulator/helper_functions.cc:
```diff
@@ -1484,6 +1484,225 @@ namespace aspect
+    MaterialModel::ReactionRateOutputs<dim> *reaction_rate_outputs_T
+      = out_T.template get_additional_output<MaterialModel::ReactionRateOutputs<dim> >();
+
+    AssertThrow(reaction_rate_outputs_C != NULL,
```
check that both are not null

tjhei commented on source/simulator/parameters.cc:
```diff
@@ -381,6 +389,27 @@ namespace aspect
+
+        prm.declare_entry ("Reaction time steps per advection step", "0",
+                           Patterns::Integer (0),
+                           "The number of reaction time steps done within one advection time step"
```
missing space at the end

tjhei commented on source/simulator/parameters.cc:
```diff
@@ -381,6 +389,27 @@ namespace aspect
+                           "The number of reaction time steps done within one advection time step"
+                           "in case operator splitting is used. This is only used if the nonlinear "
+                           "solver scheme ``operator splitting'' is selected. If set to zero, this "
+                           "parameter is ignored. Otherwise the reaction is chosen accordning to "
```
typo: according

tjhei commented on source/simulator/parameters.cc:
```diff
@@ -381,6 +389,27 @@ namespace aspect
+                           "Set a time step size for computing reactions of compositional fields and the "
+                           "temperature field in case operator splitting is used. This is only used "
+                           "when the nonlinear solver scheme ``operator splitting'' is selected. "
+                           "Units: Years or seconds, depending on the ``Use years "
```
What should I pick if I want to disable this and only use the number of steps below? 0? infinity? Document that.

tjhei replied:
also: what happens right now if it is set to 0?

jdannberg replied:
> also: what happens right now if it is set to 0?

Oh, right. You probably get a divide by zero FPE. I'll fix that, good catch!

tjhei commented on source/heating_model/interface.cc:
```diff
@@ -282,6 +283,7 @@ namespace aspect
             {
               heating_model_outputs.heating_source_terms[q] += individual_heating_outputs.heating_source_terms[q];
               heating_model_outputs.lhs_latent_heat_terms[q] += individual_heating_outputs.lhs_latent_heat_terms[q];
+              heating_model_outputs.rates_of_temperature_change[q] += individual_heating_outputs.rates_of_temperature_change[q];
```
can we assert here that the rates are 0 if operator splitting is disabled?

jdannberg commented on include/aspect/material_model/interface.h:
```diff
@@ -674,6 +674,41 @@ namespace aspect
+         * the instance of MaterialModel::Interface::evaluate() that fills
+         * the current object.
+         */
+        std::vector<std::vector<double> > reaction_rates;
```
The reaction_terms don't have the name "composition" in them either, and I think it would make sense to have similar names for those two things. I guess to me "reaction" already implied that this is for a compositional field. 

> Also: if I have a reaction that turns a composition into heat, I need to make a separate heating plugin, right?

Correct. 

> Wouldn't it make more sense if there was a temperature reaction rate in here too? Or is this not a typical case?

Hm, we could move the rates_of_temperature_change from the heating model to the material model, but it is only used in the temperature equation and nowhere else, so maybe that would be a bit confusing. On the other hand, we also have the entropy derivatives in the material model, and it's the same for them. I guess my line of thought was that a "rate of temperature change" is something more abstract than a material property, it is closer to a "heating source term" than to most of the other material model outputs, which specifically relate to a specific property of the material.  
The only case I can think of where you would want to relate changes in composition and temperature are latent heat affect, and we already have a heating model for that ("latent heat melt" already works with operator splitting). There might be other uses, but nothing that I can think of. So I don't think there would be many cases where one would have to create a new heating model. 
The advantage of splitting it is also that you can mix and match, you can use the "latent heat melt" heating model (which is quite general) with different melting models, and don't have to copy all of that stuff. And as we noticed that more and more material models are basically just an existing model with some small things changed, I think it makes a lot of sense to keep these things separate and more modular.

tjhei replied:
> So I don't think there would be many cases where one would have to create a new heating model.

Okay. One other option to consider would be to have temperature rates in the material model and the heating model. But, if you think the current design is good as is, I am happy with it. You certainly can imagine the possible use cases much better than I can.

jdannberg replied:
I think it works well the way it is right now. And if we find we need additional temperature rates in the material model, we can always add them later, that wouldn't break anything.

jdannberg commented:
That's fine, I was surprised I only got so few comments the first time :smile: 
I think I addressed all of them now.

tjhei commented:
> That's fine, I was surprised I only got so few comments the first time

At least I was still worried about the fundamental things first. You also find more stuff to nitpick about every time you review things. ;-)

tjhei commented on source/material_model/melt_simple.cc:
```diff
@@ -301,6 +311,21 @@ namespace aspect
+                  // fill reaction rate outputs if the model uses operator splitting
+                  if (this->get_parameters().use_operator_splitting)
+                    {
+                      if (reaction_rate_out != NULL)
```
so if somebody forgets to add the outputs but uses operator splitting it should not be an error?

jdannberg replied:
What do you mean by that exactly? 

Right now, everywhere in the code where we need these outputs, they are created. If you write new code where you would need the outputs, but you do not call `create_additional_named_outputs`, you get a segmentation fault. 

If someone does not implement the `create_additional_named_outputs` function in the material model, they get the error from `compute_reactions()` that the material model does not implement the ReactionRateOutputs. 

Otherwise, if we wanted to change that statement to an assert, we would have to add a call to `create_additional_named_outputs` in all places where we call evaluate, and I thought the point of having them as a separate outputs was that they don't have to be computed in all places. We do the same for the melt outputs.

tjhei commented:
After more thinking about it, I finally get it: it makes sense to only fill the rates if the additional outputs are present.

This still leaves me with the other case:
If somebody evaluates heating models with operator splitting turned on, but forgets to call create_named_outputs(), we can not evaluate the heating models correctly. Right now, you just return 0. Can you convince me that this is the correct behavior? Otherwise, I think we need an assert.

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -1497,6 +1496,31 @@ namespace aspect
+        MaterialModel::ReactionRateOutputs<dim> *reaction_rate_outputs
+          = scratch.material_model_outputs.template get_additional_output<MaterialModel::ReactionRateOutputs<dim> >();
+
+        if (reaction_rate_outputs != NULL)
```
Is this check still useful with the recent changes? If we enforce that material models only create these outputs if we use operator splitting, we would just need to ``assert( reaction_rate_outputs==NULL)`` here. Right? Sorry for the back and forth.

jdannberg replied:
Yes, makes sense.

jdannberg commented:
No, it is probably not a good behaviour to just return 0 if `create_additional_named_outputs()` is not called. You will just get the wrong solution and the problem is difficult to find. But I don't think changing this to an assert is particularly easy. I would have to find all the places in the code where we evaluate the heating model without calling `create_additional_named_outputs()` for the material model first, and there are places where evaluate is called by the manager, and places where is is called just for individual plugins, so there are many different options. And there is no easy way to test if I have found everything, because we don't have tests with operator splitting that cover all of the different postprocessors and all other options that might evaluate the heating model. 

We cannot initialize the rates_of_temperature_change to NaNs either, because then they would be NaNs for all the plugins that don't fill them, and they are all added together when the heating model manager calls the individual plugins. 

The only idea I have is that the individual heating model itself sets them to quiet_NaNs if it does not get the reaction rates. This way, if they are used anywhere without `create_additional_named_outputs` being called before, it should at least be obvious that something is wrong.

tjhei commented:
> We cannot initialize the rates_of_temperature_change to NaNs either, because then they would be NaNs for all the plugins that don't fill them, and they are all added together when the heating model manager calls the individual plugins.

I think you could add code in the manager after adding up all contributions. If ``create_additional_named_outputs`` is not given, replace the values by signaling NaNs. Then you only have to change the places that actually try to use the new rates.

jdannberg commented:
Like this? The only problem now is that someone could also call just an individual plugin (without using the heating model manager), so I would also leave my change in the latent heat melt model itself in the code (that it returns NaN if it doesn't get the additional outputs).

tjhei commented:
Yes, makes sense. But why are we getting FPEs now?

jdannberg commented:
I just put the code in the wrong line. Should be fixed now.
